### PR TITLE
Liquidation

### PR DESCRIPTION
**Pull request by kazinarka**

### CODE REVIEW

File: liquidator/src/main.rs [hunk 1]
```diff
@@ -52,52 +74,281 @@ fn main() {
...
+            self.liquidate(user);
+        }
+    }
+
+    fn liquidate(&mut self, user: &User) {
+        for queue_member in &self.oracle_data.priority_queue {
+            info!("Thread is paused for {} seconds", self.args.sleep_duration.as_secs());
+            thread::sleep(self.args.sleep_duration);
+
+            let collateral = self.collateral_data
```
Comment (tiamo) [inline liquidator/src/main.rs:183]: consider to use continue here and log error to the stdout

File: liquidator/src/main.rs [hunk 2]
```diff
@@ -52,52 +74,281 @@ fn main() {
...
+                .get(&queue_member.collateral)
+                .expect("Can't find collateral account");
+
+            let pool = self.pool_data
+                .get(&collateral.pool)
+                .expect("Can't find pool account");
```
Comment (tiamo) [inline liquidator/src/main.rs:189]: the same here

File: liquidator/src/main.rs [hunk 3]
```diff
@@ -52,52 +74,281 @@ fn main() {
...
-            info!("Withdraw request - {} in processing...", withdraw_address);
+            liquidator.process_withdraw_request(withdraw_address, withdraw_info);
+        }
+    }
+}
 
-            // TODO maybe need to get batch of users (get_user_list) outside the loop
-            let (user_key, user_data) = get_user_data(&client, withdraw_info.authority).expect("Can't get user data");
-            info!("Got data of user that made request - {}", user_key);
+pub struct Liquidator {
```
Comment (tiamo) [inline liquidator/src/main.rs:104]: move this struct to `liquidator.rs` file

File: liquidator/src/main.rs [hunk 4]
```diff
@@ -52,52 +74,281 @@ fn main() {
...
-            while amount_to_liquidate > 0 {
-                for queue_member in &oracle_data.priority_queue {
-                    // TODO send transaction
+        self.user_data = user_data.clone();
 
-                    thread::sleep(Duration::from_secs(args.sleep_duration));
+        self.user_key = get_user(withdraw_info.authority);
+        let user = user_data
+            .get(&self.user_key)
+            .expect("Can't find user account");
```
Comment (tiamo) [inline liquidator/src/main.rs:155]: It's better to return errors and no panic...

File: liquidator/src/main.rs [hunk 5]
```diff
@@ -52,52 +74,281 @@ fn main() {
...
-                        amount_to_liquidate -= queue_member.amount;
-                    }
+        if user.is_blocked {
+            warn!("User {} is blocked", user.wallet);
+            return;
+        }
+
+        self.amount_to_liquidate = withdraw_info.amount;
+
+        self.oracle_data = get_oracle_data(&self.program, self.oracle).expect("Can't get oracle account");
```
Comment (tiamo) [inline liquidator/src/main.rs:165]: the same here

File: oracle/src/main.rs [hunk 6]
```diff
@@ -58,35 +50,42 @@ fn main() {
...
     info!("Established connection: {}", args.cluster.url());
 
     // get program public key
-    let program = client.program(omnisol::id());
+    let program = client.program(id());
 
     let mut previous_queue = HashMap::new();
 
+    // find oracle PDA
+    let oracle = get_oracle();
```
Comment (tiamo) [inline oracle/src/main.rs:67]: please rename to `get_oracle_pda` or `get_oracle_address`

File: oracle/src/main.rs
Comment (tiamo) [inline oracle/src/main.rs:1]: Please check all `.expect` in the file... `expect` will stop program execution, is it really needed?

File: packages/sdk/src/client.ts [hunk 7]
```diff
@@ -42,6 +47,7 @@ export class OmnisolClient {
   static IDL = IDL
   static clock = web3.SYSVAR_CLOCK_PUBKEY
   static stakeProgram = web3.StakeProgram.programId
+  static stakingPoolProgram = new web3.PublicKey('SPoo1Ku8WFXoNDMHPsrGSTSG1Y47rzgn41SLUNakuHy')
```
Comment (tiamo) [inline packages/sdk/src/client.ts:50]: It's better to create const variable `STAKE_POOL_PROGRAM_ID`

File: packages/sdk/src/client.ts [hunk 8]
```diff
@@ -275,6 +288,79 @@ export class OmnisolClient {
...
+    const [manager] = await this.pda.manager(payer)
+    const instruction = createUpdatePoolInstruction(
+      {
+        pool: props.pool,
+        authority: payer,
+        manager,
+      },
+      {
+        data: {
+          minDeposit: props.minDeposit ?? null,
```
Comment (tiamo) [inline packages/sdk/src/client.ts:302]: is undefined raise an error here ?

File: packages/sdk/src/client.ts [hunk 9]
```diff
@@ -607,34 +728,118 @@ export class OmnisolClient {
...
+    let anchorRemainingAccounts: web3.AccountMeta[]
+    if (collateralData.isNative) {
+      const splitStake = props.splitStake ?? web3.Keypair.generate().publicKey
+      anchorRemainingAccounts = [{
+        pubkey: splitStake,
+        isWritable: true,
+        isSigner: true,
+      }]
+    } else {
+      const stakePool = props.stakePool ?? web3.Keypair.generate().publicKey
```
Comment (tiamo) [inline packages/sdk/src/client.ts:756]: No need to use `web3.Keypair.generate().publicKey` here... `stakePool` should be required prop. Please check other props

File: programs/omnisol/src/instructions/liquidate_collateral.rs [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+    if !pool.is_active {
+        return Err(ErrorCode::PoolAlreadyPaused.into());
+    }
+
+    let user = &mut ctx.accounts.user;
+    let collateral_owner = &mut ctx.accounts.collateral_owner;
+
+    if collateral_owner.is_blocked {
+        msg!("Collateral owner is blocked");
```
Comment (tiamo) [inline programs/omnisol/src/instructions/liquidate_collateral.rs:29]: no need to add return here ?

File: programs/omnisol/src/state.rs [hunk 11]
```diff
@@ -9,21 +9,45 @@ pub struct Pool {
...
     /// An account with authority that can manage and close the pool.
     pub authority: Pubkey,
-    /// Oracle address for clarification
-    pub oracle: Pubkey,
     /// Address of LP token or native stake program
-    /// TODO: rename
     pub stake_source: Pubkey,
     /// Total stake in deposit
     pub deposit_amount: u64,
+    /// Current amount of pools' collaterals
```
Comment (tiamo) [inline programs/omnisol/src/state.rs:16]: typo: of pools'

**Review by tiamo — CHANGES_REQUESTED** [review]

File: packages/sdk/src/client.ts (hunk 9 above)
Comment (kazinarka) [reply packages/sdk/src/client.ts:756]: Pubkey.default()

**Pull request merged**